### ISSUE TITLE
Multiple unquotes blanks in echo

### ISSUE THREAD

**Issue by sjvudp** [koalaman/shellcheck#377]
Maybe you want to add this: Multiple unquotes blanks in echo are not flagged, like this:

```
echo Looking around...       done
```

(Multiple blanks will be condensed to one)

**Comment by arth1** [comment]
That may be what you want.  echo is often used to strip spaces in bourne shell, like

freespace=$(echo $(df / | tail -1) | cut -d' ' -f3)

**Comment by deryni** [comment]
It may be what you want but it isn't safe in general and if you know you want it then you can tell shellcheck to not warn about it with a directive.

**Comment by brother** [comment]
@arth1 that line would render shellcheck warnings though. might be usual practice but still not clean enough.

in the case of fetching free space via df using "regular tools" I would turn to awk

df -P / | awk '/^\//{print $4}'

for the general case of using echo to clean out spaces you could consider using variable expansions.

**Comment by arth1** [comment]
Keep in mind that awk and bash aren't always available - especially in embedded.
It's more likely that sed is available, but not guaranteed, and this isn't any prettier:

freespace=$(df -P / | tail -1 | sed 's/[[:space:]][[:space:]]*/ /g' | cut -d' ' -f4)

**Comment by deryni** [comment]
Sure but again in those environments are you virtually guaranteed to be dealing with someone well versed enough to understand the problems and be able to add the shellcheck directive to silence the warning. As opposed to the naive user for whom the whitespace combining is a surprise.

**Comment by arth1** [comment]
I'm not so sure.  One bug here was due to exactly this - one user had in the past written a script that used echo to strip spaces, and another one removed a "useless echo" and added quotes because of shellcheck suggesting it, and the script's functionality broke.
Using echo to strip extra spaces might not be good programming, but it's common enough practice that I think shellcheck should at least consider whether it could be intentional.

**Comment by deryni** [comment]
Like we've been saying. It may be common but it is not safe or reliable. Which are the things shellcheck is trying to encourage. It may work for you in your specific examples but it is a bad practice and should be documented (with a directive) as such as far as I'm concerned. The number of people who have problems with this sort of thing on StackOverflow (for example) is *enormous*.

**Comment by koalaman** [comment]
I fixed the markup of the original post. It's not about word splitting during expansions but about literal whitespace between shell words.

**Comment by deryni** [comment]
You know I knew that and yet somehow didn't connect far enough to realize that was a different case entirely. =(
Sorry for the noise.

**Comment by koalaman** [comment]
A warning is now in place whenever there are 4 or more spaces between words passed to `echo` on the same line. Thanks!

Smaller gaps were not included because having two spaces was common but seemingly unintentional and harmless, and gaps between `echo` and the first word appeared to be mostly due to people padding their commands to make their arguments line up.

**Comment by sjvudp** [comment]
I just realized that "unquotes" should have read "unquoted" (hit the neighbor key) all the time, but it seems I cannot correct it.
If possible, correct that.

**Comment by sjvudp** [comment]
> A warning is now in place whenever there are 4 or more spaces between words passed to `echo` on the same line. Thanks!
> 
> Smaller gaps were not included because having two spaces was common but seemingly unintentional and harmless, and gaps between `echo` and the first word appeared to be mostly due to people padding their commands to make their arguments line up.

But `echo a  b` will be the same as `echo a b`, but I doubt that the majority deliberately aligns the arguments on echo when the output will be different.
